src/main/java/io/debezium/connector/jdbc/JdbcChangeEventSink.java

```diff
@@ -0,0 +1,406 @@
+            int index = bindKeyValuesToQuery(record, query, 1);
+            bindNonKeyValuesToQuery(record, query, index);
+
+            query.executeUpdate();
+            transaction.commit();
+        }
+        catch (Exception e) {
+            transaction.rollback();
+            throw e;
+        }
+    }
+
+    private void writeUpdate(String sql, SinkRecordDescriptor record) throws SQLException {
+        final Transaction transaction = session.beginTransaction();
+        try {
+            LOGGER.trace("SQL: {}", sql);
+            final NativeQuery<?> query = session.createNativeQuery(sql, Object.class);
+            int index = bindNonKeyValuesToQuery(record, query, 1);
+            bindKeyValuesToQuery(record, query, index);
+
+            query.executeUpdate();
+            transaction.commit();
+        }
+        catch (Exception e) {
+            transaction.rollback();
+            throw e;
+        }
+    }
+
+    private void writeDelete(String sql, SinkRecordDescriptor record) throws SQLException {
+        if (!config.isDeleteEnabled()) {
+            LOGGER.debug("Deletes are not enabled, skipping delete for topic '{}'", record.getTopicName());
+            return;
+        }
+        final Transaction transaction = session.beginTransaction();
+        try {
+            LOGGER.trace("SQL: {}", sql);
+            final NativeQuery<?> query = session.createNativeQuery(sql, Object.class);
+            bindKeyValuesToQuery(record, query, 1);
+
+            query.executeUpdate();
+            transaction.commit();
+        }
+        catch (Exception e) {
+            transaction.rollback();
+            throw e;
+        }
+    }
+
+    private int bindFieldValue(NativeQuery<?> query, FieldDescriptor field, Object value, int index) {
+        return index + dialect.bindValue(field, query, index, value);
+    }
+
+    private int bindKeyValuesToQuery(SinkRecordDescriptor record, NativeQuery<?> query, int index) {
+        if (PrimaryKeyMode.KAFKA.equals(config.getPrimaryKeyMode())) {
+            query.setParameter(index++, record.getTopicName());
+            query.setParameter(index++, record.getRawRecord().kafkaPartition());
+            query.setParameter(index++, record.getRawRecord().kafkaOffset());
+        }
+        else if (!record.getKeyFieldNames().isEmpty()) {
+            // There are key fields, either mapped to RECORD_KEY or RECORD_VALUE potentially
+            if (PrimaryKeyMode.RECORD_KEY.equals(config.getPrimaryKeyMode())) {
+                final Schema keySchema = record.getRawRecord().keySchema();
+                if (keySchema != null && Schema.Type.STRUCT.equals(keySchema.type())) {
+                    for (String fieldName : record.getKeyFieldNames()) {
+                        final FieldDescriptor field = record.getFields().get(fieldName);
+                        final Object fieldValue = ((Struct) record.getRawRecord().key()).get(fieldName);
```

Comment:
Can `SinkRecordDescriptor` provide method `getKeyFieldValue` that would behave differently based on primary key mode? Maybe the method then could be compressed to only two or even one case.